diff --git a/src/sonic-host-services/scripts/poeallocp b/src/sonic-host-services/scripts/poeallocp
new file mode 100755
index 0000000..e31b7d3
--- /dev/null
+++ b/src/sonic-host-services/scripts/poeallocp
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
+
+"""
+    poeallocp
+"""
+
+
+try:
+    import os
+    import subprocess
+    import sys
+    import time
+    import sonic_platform
+
+    from sonic_py_common import daemon_base
+    from swsscommon import swsscommon
+    from swsscommon.swsscommon import ConfigDBConnector
+except ImportError as err:
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "poeallocp"
+
+ERROR_CHASSIS_LOAD = 2
+ERROR_NOT_IMPLEMENTED = 5
+
+# Global platform-specific Chassis class instance
+platform_chassis = None
+
+# Global platform poe object instance
+platform_chassis_poe = None
+
+def load_platform_chassis():
+    global platform_chassis
+
+    # Load new platform api class
+    try:
+        platform_chassis = sonic_platform.platform.Platform().get_chassis()
+    except Exception as e:
+        log.log_error(
+            "Failed to instantiate Chassis due to {}".format(repr(e)))
+
+    if not platform_chassis:
+        return False
+
+    return True
+
+def load_chassis_poe():
+    global platform_chassis_poe
+
+    # Load platform poe
+    try:
+        platform_chassis_poe = platform_chassis.get_poe()
+    except Exception as e:
+        log.log_error(
+            "Failed to instantiate Chassis poe due to {}".format(repr(e)))
+
+    if not platform_chassis_poe:
+        return False
+
+    return True
+
+class PoeAllocpower(daemon_base.DaemonBase):
+    REDIS_TIMEOUT_MS = 0
+
+    def __init__(self, log_identifier):
+        super(PoeAllocpower, self).__init__(log_identifier)
+
+        # Open a handle to the State database
+        self.state_db = swsscommon.DBConnector("STATE_DB",
+                                              self.REDIS_TIMEOUT_MS,
+                                              False)
+        self.state_poe_table = swsscommon.Table(self.state_db, "LLDPPOE")
+
+    def platform_alloc_poe_power(self, key, poe_table_dict):
+        alloc_power = poe_table_dict.get("allocated")
+        req_power = poe_table_dict.get("requested")
+
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        poe_dict = config_db.get_table('POE')
+
+        if int(req_power) != 0 and alloc_power == '0':
+            platform_chassis_poe.set_port_lldp_pd_req(poe_dict[key]['lanes'], int(req_power))
+            time.sleep(0.25)
+            lldp_info_dict = platform_chassis_poe.get_port_lldp_pse_data(poe_dict[key]['lanes'])
+            if lldp_info_dict['lldpstatus'] == 'OK':
+                self.state_poe_table.hset(key, 'allocated', lldp_info_dict['alloc_power'])
+
+    def platform_clear_poe_power(self, key):
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        poe_dict = config_db.get_table('POE')
+
+        platform_chassis_poe.set_port_lldp_pd_req(poe_dict[key]['lanes'], 0)
+
+    def lldp_alloc_power(self, key, op, fvp):
+        if op == "SET":
+            if fvp:
+                self.platform_alloc_poe_power(key, dict(fvp))
+        elif op == "DEL":
+            if fvp:
+                self.platform_clear_poe_power(key)
+
+    def run(self):
+        """
+        Subscribes to notifications of changes in the LLDPPOE table
+        of the Redis state database.
+        """
+        self.log_info("Starting up...")
+
+        if not os.geteuid() == 0:
+            self.log_error("Must be root to run this daemon")
+            print("Error: Must be root to run this daemon")
+            sys.exit(1)
+
+        # Set select timeout to 10 seconds
+        SELECT_TIMEOUT_MS = 1000 * 10
+
+        # Daemon is paused on the configuration file to avoid lldp packets with wrong information
+        # until all interfaces are well configured on lldpd
+        resume_lldp_sent = False
+        start_time = time.time()
+
+        sel = swsscommon.Select()
+
+        # Subscribe to POE table notifications in the Config DB
+        sst_poe_statedb = swsscommon.SubscriberStateTable(self.state_db, "LLDPPOE")
+        sel.addSelectable(sst_poe_statedb)
+
+        # Listen for changes to the PORT table in the CONFIG_DB and APP_DB
+        while True:
+            (state, selectableObj) = sel.select(SELECT_TIMEOUT_MS)
+
+            if state == swsscommon.Select.OBJECT:
+                if selectableObj.getFd() == sst_poe_statedb.getFd():
+                    (key, op, fvp) = sst_poe_statedb.pop()
+                    self.lldp_alloc_power(key, op, fvp)
+                else:
+                    self.log_error("Got unexpected selectable object")
+
+# ============================= Functions =============================
+
+
+def main():
+
+  # Load platform-specific Chassis class
+    if not load_platform_chassis():
+       sys.exit(ERROR_CHASSIS_LOAD)
+
+    # Load platform poe api
+    if not load_chassis_poe():
+       sys.exit(ERROR_NOT_IMPLEMENTED)
+
+    # Instantiate a PoeAllocpower object
+    poe = PoeAllocpower(SYSLOG_IDENTIFIER)
+
+    # Log all messages from INFO level and higher
+    poe.set_min_log_priority_info()
+
+    poe.run()
+
+if __name__ == "__main__":
+    main()
diff --git a/src/sonic-host-services/setup.py b/src/sonic-host-services/setup.py
index 2f41b0d..fea14d4 100644
--- a/src/sonic-host-services/setup.py
+++ b/src/sonic-host-services/setup.py
@@ -20,6 +20,7 @@ setup(
         'scripts/procdockerstatsd',
         'scripts/determine-reboot-cause',
         'scripts/poeinit',
+        'scripts/poeallocp',
         'scripts/process-reboot-cause',
         'scripts/sonic-host-server'
     ],
