diff --git a/src/sonic-utilities/config/main.py b/src/sonic-utilities/config/main.py
old mode 100644
new mode 100755
index a5f7ca94..c7aa52a4
--- a/src/sonic-utilities/config/main.py
+++ b/src/sonic-utilities/config/main.py
@@ -1285,6 +1285,10 @@ def save(filename):
         with open(file, 'w') as config_db_file:
             json.dump(config_db, config_db_file, indent=4)

+    command = "poeutil save-conf"
+    clicommon.run_command(command)
+
+
 @config.command()
 @click.option('-y', '--yes', is_flag=True)
 @click.argument('filename', required=False)
@@ -3825,6 +3829,26 @@ def remove_neighbor(neighbor_ip_or_hostname):
     if not removed_neighbor:
         click.get_current_context().fail("Could not locate neighbor '{}'".format(neighbor_ip_or_hostname))

+
+@config.group(cls=clicommon.AbbreviationGroup)
+def poe():
+    """PoE-related configuration tasks"""
+    pass
+
+
+@poe.command()
+@click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def preemptive_priority(state):
+    command = "poeutil preemptive-priority %s" % (state)
+    clicommon.run_command(command)
+
+
+@poe.command()
+@click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def power_redundant(state):
+    command = "poeutil power-redundant %s" % (state)
+    clicommon.run_command(command)
+
 #
 # 'interface' group ('config interface ...')
 #
@@ -3842,6 +3866,206 @@ def interface(ctx, namespace):
     config_db = ConfigDBConnector(use_unix_socket_path=True, namespace=str(namespace))
     config_db.connect()
     ctx.obj = {'config_db': config_db, 'namespace': str(namespace)}
+
+
+#
+# 'poe' subcommand ('config interface poe ...')
+#
+
+@interface.group(cls=clicommon.AbbreviationGroup)
+@click.pass_context
+def poe(ctx):
+    """Config interface poe feature"""
+    pass
+
+
+@poe.command()
+@click.pass_context
+@click.argument('interface_name', metavar='<interface_name>', required=True)
+@click.argument('port_priority', metavar='<critical|high|low>', required=True, type=click.Choice(['critical', 'high', 'low']))
+def priority(ctx, interface_name, port_priority):
+    """Config interface PoE priority"""
+    # Get the config_db connector
+    config_db = ctx.obj['config_db']
+
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) > 1 and multi_asic.is_multi_asic():
+        ctx.fail("Interface range not supported in multi-asic platforms !!")
+
+    if len(intf_fs) == 1 and interface_name_is_valid(config_db, interface_name) is False:
+        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+
+    # cmd = 'poeutil'
+    # args = ['priority', interface_name, port_priority]
+    command = "poeutil port-priority %s %s" % (interface_name, port_priority)
+    clicommon.run_command(command)
+
+def power_inline_p(ctx, interface_name, power_mode, max_power_float):
+    """Config interface PoE power"""
+    # Get the config_db connector
+    config_db = ctx.obj['config_db']
+
+    max_power = int(float(max_power_float) * 10)  # Transfer to 0.1W
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) > 1 and multi_asic.is_multi_asic():
+        ctx.fail("Interface range not supported in multi-asic platforms !!")
+
+    if len(intf_fs) == 1 and interface_name_is_valid(config_db, interface_name) is False:
+        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+
+    command = "poeutil power-inline %s %s %s" % (
+        interface_name, power_mode, max_power)
+
+    # print(command)
+    clicommon.run_command(command)
+
+@ poe.command()
+@ click.pass_context
+@ click.argument('interface_name', metavar='<interface_name>', required=True)
+@ click.argument('power_mode', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+@ click.argument('max_power_float', metavar='<max_power_float>', required=True, type=click.FloatRange(0, 90))
+def power_inline(ctx, interface_name, power_mode, max_power_float):
+    """Config interface PoE power"""
+    # Get the config_db connector
+    config_db = ctx.obj['config_db']
+    poe_dict = config_db.get_table('POE')
+
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) > 1 and multi_asic.is_multi_asic():
+        ctx.fail("Interface range not supported in multi-asic platforms !!")
+
+    if len(intf_fs) == 1 and interface_name_is_valid(config_db, interface_name) is False:
+        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+
+    # power_mode options
+    # 0 - disable with dynamic power calculation (0b00)
+    # 1 - enable with dynamic power calculation  (0b01)
+    # 2 - disable with static power calculation  (0b10)
+    # 3 - enable with static power calculation   (0b11)
+
+    # bit 2 is power_calculation mode
+    # dynamic mode : bit2 is 0
+    # static mode : bit 2 is 1
+    power_calculation = "static" if (poe_dict[interface_name]["power_mode"] == "disableStatic" or
+                         poe_dict[interface_name]["power_mode"] == "enableStatic" or
+                         poe_dict[interface_name]["power_mode"] == "disable") else "dynamic"
+
+    if power_calculation == "static": # static
+        if power_mode == "enable":
+            power_mode_p = "enableStatic"
+        else:
+            power_mode_p = "disableStatic"
+    else:
+        if power_mode == "enable":
+            power_mode_p = "enableDynamic"
+        else:
+            power_mode_p = "disableDynamic"
+
+    power_inline_p(ctx, interface_name, power_mode_p, max_power_float)
+
+@ poe.command()
+@ click.pass_context
+@ click.argument('interface_name', metavar='<interface_name>', required=True)
+@ click.argument('power_calculation', metavar='<static|dynamic>', required=True, type=click.Choice(['static', 'dynamic']))
+def power_calculation(ctx, interface_name, power_calculation):
+    """Config interface PoE power"""
+    # Get the config_db connector
+    config_db = ctx.obj['config_db']
+    poe_dict = config_db.get_table('POE')
+
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) > 1 and multi_asic.is_multi_asic():
+        ctx.fail("Interface range not supported in multi-asic platforms !!")
+
+    if len(intf_fs) == 1 and interface_name_is_valid(config_db, interface_name) is False:
+        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+
+    # power_mode options
+    # 0 - disable with dynamic power calculation (0b00)
+    # 1 - enable with dynamic power calculation  (0b01)
+    # 2 - disable with static power calculation  (0b10)
+    # 3 - enable with static power calculation   (0b11)
+
+    # bit 1 is power_status mode
+    # disable mode : bit 1 is 0
+    # enable mode : bit 1  is 1
+    power_status = "enable" if (poe_dict[interface_name]["power_mode"] == "enableDynamic" or
+                         poe_dict[interface_name]["power_mode"] == "enableStatic" or
+                         poe_dict[interface_name]["power_mode"] == "enable") else "disable"
+    #print(port_dict[interface_name]["poe_status"])
+    #print("power_status:", power_status)
+    if power_status == "enable": # enable
+        if power_calculation == "static":
+            power_mode_p = "enableStatic"
+        else:
+            power_mode_p = "enableDynamic"
+    else:
+        if power_calculation == "dynamic":
+            power_mode_p = "disableDynamic"
+        else:
+            power_mode_p = "disableStatic"
+
+    max_power = poe_dict[interface_name]["maxpower"]
+
+    power_inline_p(ctx, interface_name, power_mode_p, max_power)
+
+
+@ interface.group(cls=clicommon.AbbreviationGroup)
+@ click.pass_context
+def lldp(ctx):
+    """Configuring interface LLDP feature"""
+    pass
+
+
+@ lldp.command('mdi-tlv')
+@ click.pass_context
+@ click.argument('interface_name', metavar='<interface_name>', required=True)
+@ click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def mdi_tlv(ctx, interface_name, state):
+    """Config interface LLDP mdi-tlv"""
+    # Get the config_db connector
+    config_db = ctx.obj['config_db']
+
+    if clicommon.get_interface_naming_mode() == "alias":
+        interface_name = interface_alias_to_name(config_db, interface_name)
+        if interface_name is None:
+            ctx.fail("'interface_name' is None!")
+
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) > 1 and multi_asic.is_multi_asic():
+        ctx.fail("Interface range not supported in multi-asic platforms !!")
+
+    if len(intf_fs) == 1 and interface_name_is_valid(config_db, interface_name) is False:
+        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+
+    click.echo("{} LLDP MDI feature.".format(
+        "Enable" if state == "enable" else "Disable"))
+
+    command = "poeutil mdi-tlv %s %s" % (
+        interface_name, state)
+    clicommon.run_command(command)
+
+
 #
 # 'startup' subcommand
 #
diff --git a/src/sonic-utilities/poeutil/__init__.py b/src/sonic-utilities/poeutil/__init__.py
new file mode 100644
index 00000000..e69de29b
diff --git a/src/sonic-utilities/poeutil/main.py b/src/sonic-utilities/poeutil/main.py
new file mode 100755
index 00000000..58adeb7c
--- /dev/null
+++ b/src/sonic-utilities/poeutil/main.py
@@ -0,0 +1,572 @@
+#!/usr/bin/env python3:
+
+import os
+import sys
+import click
+import subprocess
+import sonic_platform
+import sonic_platform_base.sonic_sfp.sfputilhelper
+import time
+import shutil
+
+from sonic_platform_base.sfp_base import SfpBase
+from sonic_py_common import device_info, logger, multi_asic
+from swsscommon.swsscommon import ConfigDBConnector
+from utilities_common.netstat import ns_diff, table_as_json, format_brate, format_prate, format_util, format_number_with_comma
+from tabulate import tabulate
+from utilities_common.intf_filter import parse_interface_in_filter
+import re
+
+SYSLOG_IDENTIFIER = "poeutil"
+
+EXIT_FAIL = -1
+EXIT_SUCCESS = 0
+ERROR_PERMISSIONS = 1
+ERROR_CHASSIS_LOAD = 2
+ERROR_SFPUTILHELPER_LOAD = 3
+ERROR_PORT_CONFIG_LOAD = 4
+ERROR_NOT_IMPLEMENTED = 5
+ERROR_INVALID_PORT = 6
+SMBUS_BLOCK_WRITE_SIZE = 32
+
+# Global platform-specific Chassis class instance
+platform_chassis = None
+
+# Global platform-specific sfputil class instance
+platform_sfputil = None
+
+# Global platform poe object instance
+platform_chassis_poe = None
+
+# Global logger instance
+log = logger.Logger(SYSLOG_IDENTIFIER)
+
+def atoi(text):
+    return int(text) if text.isdigit() else text
+
+def natural_keys(text):
+    '''
+    alist.sort(key=natural_keys) sorts in human order
+    http://nedbatchelder.com/blog/200712/human_sorting.html
+    (See Toothy's implementation in the comments)
+    '''
+    return [ atoi(c) for c in re.split(r'(\d+)', text) ]
+
+def is_port_type_poe(poe_class):
+    if (poe_class == "NA"):
+        return False
+    else:
+        return True
+
+
+# Instantiate platform-specific Chassis class
+def load_platform_chassis():
+    global platform_chassis
+
+    # Load new platform api class
+    try:
+        platform_chassis = sonic_platform.platform.Platform().get_chassis()
+    except Exception as e:
+        log.log_error(
+            "Failed to instantiate Chassis due to {}".format(repr(e)))
+
+    if not platform_chassis:
+        return False
+
+    return True
+
+
+# Instantiate SfpUtilHelper class
+def load_sfputilhelper():
+    global platform_sfputil
+
+    # we have to make use of sfputil for some features
+    # even though when new platform api is used for all vendors.
+    # in this sense, we treat it as a part of new platform api.
+    # we have already moved sfputil to sonic_platform_base
+    # which is the root of new platform api.
+    platform_sfputil = sonic_platform_base.sonic_sfp.sfputilhelper.SfpUtilHelper()
+
+    if not platform_sfputil:
+        return False
+
+    return True
+
+
+def load_port_config():
+    try:
+        if multi_asic.is_multi_asic():
+            # For multi ASIC platforms we pass DIR of port_config_file_path and the number of asics
+            (platform_path, hwsku_path) = device_info.get_paths_to_platform_and_hwsku_dirs()
+
+            # Load platform module from source
+            platform_sfputil.read_all_porttab_mappings(
+                hwsku_path, multi_asic.get_num_asics())
+        else:
+            # For single ASIC platforms we pass port_config_file_path and the asic_inst as 0
+            port_config_file_path = device_info.get_path_to_port_config_file()
+            platform_sfputil.read_porttab_mappings(port_config_file_path, 0)
+    except Exception as e:
+        log.log_error("Error reading port info ({})".format(str(e)), True)
+        return False
+
+    return True
+
+def load_chassis_poe():
+    global platform_chassis_poe
+
+    # Load platform poe
+    try:
+        platform_chassis_poe = platform_chassis.get_poe()
+    except Exception as e:
+        log.log_error(
+            "Failed to instantiate Chassis poe due to {}".format(repr(e)))
+
+    if not platform_chassis_poe:
+        return False
+
+    return True
+
+def get_physical_port_name(logical_port, physical_port, ganged):
+    """
+        Returns:
+          port_num if physical
+          logical_port:port_num if logical port and is a ganged port
+          logical_port if logical and not ganged
+    """
+    if logical_port == physical_port:
+        return str(logical_port)
+    elif ganged:
+        return "{}:{} (ganged)".format(logical_port, physical_port)
+    else:
+        return logical_port
+
+
+def logical_port_name_to_physical_port_list(port_name):
+    if port_name.startswith("Ethernet"):
+        if platform_sfputil.is_logical_port(port_name):
+            return platform_sfputil.get_logical_to_physical(port_name)
+        else:
+            click.echo("Error: Invalid port '{}'".format(port_name))
+            return None
+    else:
+        return [int(port_name)]
+
+
+def print_all_valid_port_values():
+    click.echo("Valid values for port: {}\n".format(
+        str(platform_sfputil.logical)))
+
+
+def run_lldpcli_config_port_command(interface, enabled):
+    # lldpcli configure port Etherent40 dot3 power pse supported enabled powerpairs signal class class-4 type 2  source primary priority low requested 28000 allocated 15000
+
+    cmd = "lldpcli configure port %s dot3 power pse supported %s powerpairs signal class class-4" % (
+        interface, enabled)
+    # print(cmd)
+    try:
+        result = subprocess.check_output(
+            cmd, shell=True, stderr=subprocess.STDOUT, universal_newlines=True)
+        # print(result.strip())
+        return result.strip()
+    except subprocess.CalledProcessError as e:
+        click.echo("Error running run_lldpcli_neighbor_command command:", e)
+        return None
+
+
+# This is our main entrypoint - the main 'poeutil' command
+@click.group()
+def cli():
+    """poeutil - Command line utility for managing PoE devices"""
+
+    # Load platform-specific Chassis class
+    if not load_platform_chassis():
+        sys.exit(ERROR_CHASSIS_LOAD)
+
+    # Load SfpUtilHelper class
+    if not load_sfputilhelper():
+        sys.exit(ERROR_SFPUTILHELPER_LOAD)
+
+    # Load port info
+    if not load_port_config():
+        sys.exit(ERROR_PORT_CONFIG_LOAD)
+
+    # Load port info
+    if not load_chassis_poe():
+        sys.exit(ERROR_NOT_IMPLEMENTED)
+
+
+# 'save_conf' subcommand
+@cli.command()
+def save_conf():
+    """ Save PoE Configure"""
+    # As the PoE configuration database is designed as part of port configuration
+    # in cnfig_db.json, we just save the configuration in the PoE controller and
+    # the PoE global configurationhere
+
+    # click.echo("Save PoE configuartion ... ", nl=False)
+    platform_chassis_poe.save_config()
+
+
+@cli.command()
+@click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def preemptive_priority(state):
+    """Configure the PoE priority ignorance"""
+
+    click.echo("{} PoE preemptive priority  ... ".format(
+        "Enabling" if state == 'enable' else "Disabling"), nl=False)
+
+    result = platform_chassis_poe.set_preemptive_priority(state)
+
+    if result:
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        config_db.mod_entry("POE", 'Global', {'preemptive_priority': state})
+        click.echo("OK")
+    else:
+        click.echo("Failed")
+
+
+@cli.command()
+@click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def power_redundant(state):
+    """Configure the PoE power redundant"""
+
+    click.echo("{} PoE power redundance  ... ".format(
+        "Enabling" if state == 'enable' else "Disabling"), nl=False)
+
+    result = platform_chassis_poe.set_power_redundant(state)
+
+    if result:
+        config_db = ConfigDBConnector()
+        config_db.connect()
+        config_db.mod_entry("POE", 'Global', {'power_redundant': state})
+        click.echo("OK")
+    else:
+        click.echo("Failed")
+
+
+@ cli.command()
+@ click.argument('port_name', metavar='<port_name>')
+@ click.argument('state', metavar='<enable|disable>', required=True, type=click.Choice(['enable', 'disable']))
+def mdi_tlv(port_name, state):
+    """Configure the LLDP of PoE Ports"""
+    ganged = False
+    i = 1
+
+    if platform_sfputil.is_logical_port(port_name) == 0:
+        click.echo("Error: invalid port '{}'\n".format(port_name))
+        print_all_valid_port_values()
+        sys.exit(ERROR_INVALID_PORT)
+
+    physical_port_list = logical_port_name_to_physical_port_list(port_name)
+    if physical_port_list is None:
+        click.echo(
+            "Error: No physical ports found for logical port '{}'".format(port_name))
+        return
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_port = config_db.get_entry('POE', port_name)
+
+    if not is_port_type_poe(poe_port['class']):
+        click.echo(
+            "LLDP MDI-TLV is not applicable for none-PoE port {}.".format(port_name))
+        sys.exit(EXIT_FAIL)
+
+    if len(physical_port_list) > 1:
+        ganged = True
+
+    for physical_port in physical_port_list:
+        click.echo("{} LLDP MDI-TLV for port {} ... ".format("Enabling" if state == 'enable' else "Disabling",
+                                                             get_physical_port_name(port_name, i, ganged)), nl=False)
+
+        # Update lldpd dot3 status
+        if (state == 'enable'):
+            enabled = 'enabled'
+        else:
+            enabled = ''
+        run_lldpcli_config_port_command(port_name, enabled)
+
+        # Save to DB
+        config_db.mod_entry("POE", port_name, {'lldp': state})
+        click.echo("OK")
+
+        i += 1
+
+
+@ cli.command()
+@ click.argument('port_name', metavar='<port_name>')
+@ click.argument('priority', metavar='<critical|high|low>', required=True, type=click.Choice(['critical', 'high', 'low']))
+def port_priority(port_name, priority):
+    """Configure the priority of PoE Ports"""
+    ganged = False
+    i = 1
+
+    if platform_sfputil.is_logical_port(port_name) == 0:
+        click.echo("Error: invalid port '{}'\n".format(port_name))
+        print_all_valid_port_values()
+        sys.exit(ERROR_INVALID_PORT)
+
+    physical_port_list = logical_port_name_to_physical_port_list(port_name)
+    # click.echo("physical_port_list '{}'\n".format(physical_port_list))
+    if physical_port_list is None:
+        click.echo(
+            "Error: No physical ports found for logical port '{}'".format(port_name))
+        return
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_port = config_db.get_entry('POE', port_name)
+
+    if not is_port_type_poe(poe_port['class']):
+        click.echo(
+            "PoE priority is not applicable for none-PoE port {}.".format(port_name))
+        sys.exit(EXIT_FAIL)
+
+    if len(physical_port_list) > 1:
+        ganged = True
+
+    for physical_port in physical_port_list:
+        click.echo("Set port priority for port {} ... ".format(
+            get_physical_port_name(port_name, i, ganged)), nl=False)
+
+        poe_port = config_db.get_entry('POE', port_name)
+
+        result = platform_chassis_poe.set_port_priority(int(poe_port['lanes']), priority)
+
+        if result:
+            config_db.mod_entry("POE", port_name, {'priority': priority})
+            click.echo("OK")
+        else:
+            click.echo("Failed")
+
+        i += 1
+
+
+@ cli.command()
+@ click.argument('port_name', metavar='<port_name>')
+@ click.argument('power_mode', metavar='<enableDynamic|enableStatic|disableDynamic|disableStatic>', required=True, type=click.Choice(['enableDynamic', 'enableStatic', 'disableDynamic', 'disableStatic']))
+@ click.argument('max_power', metavar='<max_power>', required=True, type=click.IntRange(0, 900))
+def power_inline(port_name, power_mode, max_power):
+    """Configure the power-inline of PoE Ports"""
+    ganged = False
+    i = 1
+
+    if platform_sfputil.is_logical_port(port_name) == 0:
+        click.echo("Error: invalid port '{}'\n".format(port_name))
+        print_all_valid_port_values()
+        sys.exit(ERROR_INVALID_PORT)
+
+    physical_port_list = logical_port_name_to_physical_port_list(port_name)
+    if physical_port_list is None:
+        click.echo(
+            "Error: No physical ports found for logical port '{}'".format(port_name))
+        return
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_port = config_db.get_entry('POE', port_name)
+
+    if not is_port_type_poe(poe_port['class']):
+        click.echo(
+            "power-inline is not applicable for none-PoE port {}.".format(port_name))
+        sys.exit(EXIT_FAIL)
+
+    if len(physical_port_list) > 1:
+        ganged = True
+
+    for physical_port in physical_port_list:
+        click.echo("Set port mode and max power for port {} ... ".format(
+            get_physical_port_name(port_name, i, ganged)), nl=False)
+
+        poe_port = config_db.get_entry('POE', port_name)
+        result = platform_chassis_poe.set_port_power(int(poe_port['lanes']), poe_port['class'], power_mode, float(max_power)/10)
+
+        if result:
+            config_db.mod_entry("POE", port_name, {'power_mode': power_mode})
+            config_db.mod_entry("POE", port_name, {
+                               'maxpower': float(max_power/10)})
+            click.echo("OK")
+        else:
+            click.echo("Failed")
+
+        i += 1
+
+# 'show' subgroup
+@ cli.group()
+def show():
+    """Display status of PoE Ports"""
+    pass
+
+# 'power_inline' subcommand
+
+
+@ show.command()
+@ click.option('-p', '--port', metavar='<port_name>', help="Display PoE port status for port <port_name> only")
+def power_inline(port):
+    """Display power-inline status of PoE port(s)"""
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_dict = config_db.get_table('POE')
+
+    table = []
+    header = ["Interface", "Mode", "Max-power(W)",
+              "Priority", "State", "Used(W)"]
+    # print(port_dict)
+
+    for key in sorted(poe_dict, key=natural_keys):
+        if key == 'Global': #Global setting
+            continue  # Skips
+        #
+        if poe_dict[key]["priority"] == "NA":
+            continue  # Skips the non-poe ports
+
+        if port is not None and key != port:
+            continue  # Skips the non-selected ports
+
+        max_power = poe_dict[key]["maxpower"]
+        if float(max_power) == 0.0:
+            max_power = "auto"
+
+        ps_info = platform_chassis_poe.get_port_status(int(poe_dict[key]['lanes']))
+
+        table.append((key, poe_dict[key]["power_mode"], max_power,
+                     poe_dict[key]["priority"], ps_info['portstatus'], ps_info['measuredPortPower']))
+
+    print(tabulate(table, header, tablefmt='simple', stralign='left'))
+
+
+@ show.command()
+@ click.option('-p', '--port', metavar='<port_name>', help="Display PoE port status for port <port_name> only")
+def power_status(port):
+    """Display power-inline status of PoE port(s)"""
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_dict = config_db.get_table('POE')
+
+    table = []
+    header = ["Interface", "State",
+              "Class(Pri)", "Class(Sec)", "Used(W)"]
+
+    for key in sorted(poe_dict, key=natural_keys):
+        if key == 'Global': #Global setting
+            continue  # Skips
+        #
+        if poe_dict[key]["priority"] == "NA":
+            continue  # Skips the non-poe ports
+
+        if port is not None and key != port:
+            continue  # Skips the non-selected ports
+
+        ps_info = platform_chassis_poe.get_port_status(int(poe_dict[key]['lanes']))
+
+        table.append((key, ps_info['assignedClassPrimary'], ps_info['assignedClassSecondary'],
+                     ps_info['portstatus'], ps_info['measuredPortPower']))
+
+    print(tabulate(table, header, tablefmt='simple', stralign='left'))
+
+
+@ show.command()
+@ click.option('-p', '--port', metavar='<port_name>', help="Display PoE port status for port <port_name> only")
+def power_method(port):
+    """Display power-inline status of PoE port(s)"""
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_dict = config_db.get_table('POE')
+
+    table = []
+    header = ["Interface", "Method"]
+
+    for key in sorted(poe_dict, key=natural_keys):
+        if key == 'Global': #Global setting
+            continue  # Skips
+        if poe_dict[key]["priority"] == "NA":
+            continue  # Skips the non-poe ports
+
+        if port is not None and key != port:
+            continue  # Skips the non-selected ports
+
+        lldp_pse_info = platform_chassis_poe.get_port_lldp_pse_data(int(poe_dict[key]['lanes']))
+        table.append((key, lldp_pse_info['layer2Usage']))
+
+    print(tabulate(table, header, tablefmt='simple', stralign='left'))
+
+
+@ show.command()
+@ click.option('-p', '--port', metavar='<port_name>', help="Display PoE port status for port <port_name> only")
+def power_conf(port):
+    """Display power-inline status of PoE port(s)"""
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+    poe_dict = config_db.get_table('POE')
+
+    table = []
+    header = ["Interface", "Mode", "Max-power(W)",
+              "Priority", "MDI-TLV"]
+
+    for key in sorted(poe_dict, key=natural_keys):
+        if key == 'Global': #Global setting
+            continue  # Skips
+        if poe_dict[key]["priority"] == "NA":
+            continue  # Skips the non-poe ports
+
+        if port is not None and key != port:
+            continue  # Skips the non-selected ports
+
+        max_power = poe_dict[key]["maxpower"]
+        if float(max_power) == 0.0:
+            max_power = "auto"
+
+        table.append((key, poe_dict[key]["power_mode"], max_power,
+                     poe_dict[key]["priority"], poe_dict[key]["lldp"]))
+
+    print(tabulate(table, header, tablefmt='simple', stralign='left'))
+
+
+# 'system_power' subcommand
+@ show.command()
+def system_power():
+    """Display system-power status of PoE"""
+
+    total_power_info_dict = platform_chassis_poe.get_total_power()
+
+    header = ['name', 'power(W)']
+    table = []
+
+    display_word = "Power Consumption"
+    table.append((display_word, total_power_info_dict['powerconsumption']))
+
+    display_word = "Calculated Power"
+    table.append((display_word, total_power_info_dict['calcpower']))
+
+    display_word = "Available Power"
+    table.append((display_word, total_power_info_dict['availablepower']))
+
+    display_word = "Power Limit"
+    table.append((display_word, total_power_info_dict['powerlimit']))
+
+    display_word = "Power Bank"
+    table.append((display_word, hex(int(total_power_info_dict['powerbank']))))
+
+    config_db = ConfigDBConnector()
+    config_db.connect()
+
+    poe_global = config_db.get_entry('POE', 'Global')
+    value = poe_global['preemptive_priority']
+    display_word = "Preemptive Priority"
+    table.append((display_word, value))
+
+    value = poe_global['power_redundant']
+    display_word = "Power Redundance"
+    table.append((display_word, value))
+
+    click.echo(tabulate(table, header, tablefmt='simple', stralign='left'))
+
+
+if __name__ == '__main__':
+    cli()
diff --git a/src/sonic-utilities/setup.py b/src/sonic-utilities/setup.py
index 1c4f0ae1..a1e77d85 100644
--- a/src/sonic-utilities/setup.py
+++ b/src/sonic-utilities/setup.py
@@ -62,6 +62,7 @@ setup(
         'utilities_common',
         'watchdogutil',
         'sonic_cli_gen',
+        'poeutil',
     ],
     package_data={
         'generic_config_updater': ['generic_config_updater.conf.json'],
@@ -191,6 +192,7 @@ setup(
             'undebug = undebug.main:cli',
             'watchdogutil = watchdogutil.main:watchdogutil',
             'sonic-cli-gen = sonic_cli_gen.main:cli',
+            'poeutil = poeutil.main:cli',
         ]
     },
     install_requires=[
diff --git a/src/sonic-utilities/show/interfaces/__init__.py b/src/sonic-utilities/show/interfaces/__init__.py
index a7a56244..0107e62a 100644
--- a/src/sonic-utilities/show/interfaces/__init__.py
+++ b/src/sonic-utilities/show/interfaces/__init__.py
@@ -412,6 +412,94 @@ def mpls(ctx, interfacename, namespace, display):

 interfaces.add_command(portchannel.portchannel)

+
+#
+# poe group (show interfaces trasceiver ...)
+#
+@interfaces.group(cls=clicommon.AliasedGroup)
+def poe():
+    """Show PoE port information"""
+    pass
+
+@poe.command()
+@click.argument('interfacename', required=False)
+@click.option('--namespace', '-n', 'namespace', default=None, show_default=True,
+              type=click.Choice(multi_asic_util.multi_asic_ns_choices()), help='Namespace name or all')
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def power_inline(interfacename, namespace, verbose):
+    """Show interface PoE information"""
+
+    ctx = click.get_current_context()
+
+    cmd = "poeutil show power-inline"
+
+    if interfacename is not None:
+        interfacename = try_convert_interfacename_from_alias(ctx, interfacename)
+        cmd = "poeutil show power-inline -p %s" % (interfacename)
+
+    clicommon.run_command(cmd, display_cmd=verbose)
+
+
+@poe.command()
+@click.argument('interfacename', required=False)
+@click.option('--namespace', '-n', 'namespace', default=None, show_default=True,
+              type=click.Choice(multi_asic_util.multi_asic_ns_choices()), help='Namespace name or all')
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def power_conf(interfacename, namespace, verbose):
+    """Show interface PoE information"""
+
+    ctx = click.get_current_context()
+
+    cmd = "poeutil show power-conf"
+
+    if interfacename is not None:
+        interfacename = try_convert_interfacename_from_alias(
+            ctx, interfacename)
+        cmd = "poeutil show power-conf -p %s" % (interfacename)
+
+    clicommon.run_command(cmd, display_cmd=verbose)
+
+
+@poe.command()
+@click.argument('interfacename', required=False)
+@click.option('--namespace', '-n', 'namespace', default=None, show_default=True,
+              type=click.Choice(multi_asic_util.multi_asic_ns_choices()), help='Namespace name or all')
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def power_status(interfacename, namespace, verbose):
+    """Show interface PoE information"""
+
+    ctx = click.get_current_context()
+
+    cmd = "poeutil show power-status"
+
+    if interfacename is not None:
+        interfacename = try_convert_interfacename_from_alias(
+            ctx, interfacename)
+        cmd = "poeutil show power-status -p %s" % (interfacename)
+
+    clicommon.run_command(cmd, display_cmd=verbose)
+
+
+@poe.command()
+@click.argument('interfacename', required=False)
+@click.option('--namespace', '-n', 'namespace', default=None, show_default=True,
+              type=click.Choice(multi_asic_util.multi_asic_ns_choices()), help='Namespace name or all')
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def power_method(interfacename, namespace, verbose):
+    """Show interface PoE information"""
+
+    ctx = click.get_current_context()
+
+    cmd = "poeutil show power-method"
+
+    if interfacename is not None:
+        interfacename = try_convert_interfacename_from_alias(
+            ctx, interfacename)
+        cmd = "poeutil show power-method -p %s" % (interfacename)
+
+    clicommon.run_command(cmd, display_cmd=verbose)
+
+
 #
 # transceiver group (show interfaces trasceiver ...)
 #
diff --git a/src/sonic-utilities/show/main.py b/src/sonic-utilities/show/main.py
old mode 100755
new mode 100644
index adbbad12..4e4a55e2
--- a/src/sonic-utilities/show/main.py
+++ b/src/sonic-utilities/show/main.py
@@ -392,6 +392,20 @@ def arp(ipaddress, iface, verbose):

     run_command(cmd, display_cmd=verbose)

+
+#
+# 'poe-power' command ("show poe-power")
+#
+
+@cli.command()
+@click.option('--verbose', is_flag=True, help="Enable verbose output")
+def poe_power(verbose):
+    """Show PoE status"""
+    cmd = "poeutil show system-power"
+
+    run_command(cmd, display_cmd=verbose)
+
+
 #
 # 'ndp' command ("show ndp")
 #
